step-9-selenium-tests.md

```diff
@@ -1,91 +1,91 @@
+
+        browser = new BrowserWebDriverContainer<>()
+                .withCapabilities(new ChromeOptions());
+        browser.start();
     }
 
     @Test
     public void seleniumTest() {
         RemoteWebDriver webDriver = browser.getWebDriver();
 
-        webDriver.get("http://sut:8080/view/ratings/");
+        webDriver.get("http://" + /* TODO: replace me */ + "/view/ratings/");
         WebElement content = webDriver.findElementById("content");
 
-        assert content.getText().equals("Foobar");
+        assertEquals(content.getText(), "Foobar");
     }
 
+    @After
+    public void tearDown() {
+        browser.stop();
+    }
 }
 ```
 
+## Component structure
+
+With the addition of this test code on top of `AbstractIntegrationTest`,we now have the following arrangement of components in our test:
+
+
+![Component structure diagram](https://kroki.io/graphviz/svg/eNp9UU1rAkEMvfdXhL1WD_VaVBBBUISitheRMjvGdXA6kSTbHor_vbt1dUddzW1e8j6SWbuMzX4LI_h9gqIkT4-A9bko8udL1SjLmxR9Nxl_TJPXM5jM9-xCBgMihTnyt7MYt8fvwSkoikoFH5qdOrdOi4JlKahxAVlgSHaHDDUS-byRaMYoNTIxm52pnzNcO4ljo8fg8q_nAdOPxLIXMRvXg3bvyvDuWJTi7kwU7eJcZe9RTlhWdyohJp-sTkYPOP-aTUGWljxxN_U5tk4f0D9L3uZq1DjyRA2rFNTDH-PTt5w=)
+<!--
+Regenerate if necessary at https://kroki.io/
+Source:
+
+digraph G {
+    subgraph cluster_1 {
+        label="JVM";
+        "Spring Boot Service";
+        "JUnit tests";
+    }
+    subgraph cluster_2 {
+        label="Testcontainers Docker containers";
+        Postgres;
+        Kafka;
+        Redis;
+        "Selenium+Browser container";
+    }
+    "Spring Boot Service" -> Postgres;
+    "Spring Boot Service" -> Kafka;
+    "Spring Boot Service" -> Redis;
+    "JUnit tests" -> "Selenium+Browser container" [label="control"];
+    "Selenium+Browser container" -> "Spring Boot Service" [color=blue, label="?"];
+    "JUnit tests" -> "Spring Boot Service" [label=starts];
+}
+-->
+
+## Talking to our service
+
+One final piece of the puzzle is missing, though: how can we get our dockerized web browser to communicate with the Spring Boot server running inside our local JVM? (the blue line in the diagram).
+
+Normally Docker containers cannot access the host machine, but Testcontainers offers a mechanism to expose host ports.
+Have a look at the documentation, and try and find out:
+
+* how to expose a host port to containers
+* once a host port is exposed, what URL should we tell the browser to access? (hint: it's not `localhost`)
+
+With this information, update the TODOs in the test class and see what happens.
+
+## Extra credit
+
+* The `BrowserWebDriverContainer` has a nice feature that allows test runs to be saved to a video, either for all tests or just failing tests. Check the documentation, and give this a try.
+* `BrowserWebDriverContainer` exposes a `getVncAddress()` method. Can you use this on your local machine to watch tests as they run via VNC? Even better, try and use the browser's inspect tools while stopped on a breakpoint in your tests!
```

Comment:
I wonder if it makes sense to mention the approach from [my article](https://bsideup.github.io/posts/debugging_containers/) to run it on the same port

Reply:
I think it's a bit to much for the workshop, but we could just add a link for reference.

Reply:
Agreed with both, let's add a link 😄

Reply:
Added